dra-evolution/pkg/api/capacity_types.go

```diff
@@ -3,168 +3,115 @@ package api
 import (
 	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/types"
 )
 
-const (
-	DevMgmtAPIVersion = "devmgmtproto.k8s.io/v1alpha1"
-)
-
-// DevicePool represents a collection of devices managed by a given driver. How
+// ResourcePool represents a collection of devices managed by a given driver. How
 // devices are divided into pools is driver-specific, but typically the
 // expectation would a be a pool per identical collection of devices, per node.
 // It is fine to have more than one pool for a given node, for the same driver.
-type DevicePool struct {
-	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata,omitempty"`
-
-	Spec   DevicePoolSpec   `json:"spec,omitempty"`
-	Status DevicePoolStatus `json:"status,omitempty"`
-}
-
-// DevicePoolSpec identifies the driver and contains the data for the pool
-// prior to any allocations.
-// NOTE: It's not clear that spec/status is the right model for this data.
-type DevicePoolSpec struct {
-	// NodeName is the name of the node containing the devices in the pool.
-	// For network attached devices, this may be empty.
+//
+// Where a device gets published may change over time. The unique identifier
+// for a device is the tuple `<driver name>/<node name>/<device name>`. Each
+// of these names is a DNS label or domain, so it is okay to concatenate them
+// like this in a string with a slash as separator.
```

Comment:
Interesting question. I think the scheduler actually shouldn't care; it should though be prepared for the situation. The specific pool it is in is not relevant to scheduling, particularly because we do not store allocations in the pool, and the identifier here is not dependent on the pool. However, if we are asking for more than one device, we better not accidentally pick the same device twice, because it is in different pools.

We should be able to look at the metadata and/or resourceVersion (didn't we decide to say they must be comparable now?) to choose the most recent one. That's important for the status which may show device health.

Reply:
> What should a consumer do if they find the same device in more than one pool (e.g. the driver decided to rebalance pools for size)?

At the moment, that could be treated as "one such device is available", although it's bordering on an error because it violates the uniqueness requirement. If the producer needs to rebalance, they should first remove from or delete the old pool, then add to or create another.

Is ordering of events through an informer guaranteed? I wouldn't count on it, I'm just curious.

This becomes more interesting once we use the status for allocation. That will limit how much the producer can still shuffle devices around and the consumer needs to be sure that there is no ambiguity about which pool it should update, if it needs to.

Reply:
I added some comments regarding this below.

Reply:
It's unclear whether RV or creationTimestamp is better, but we just need a clearly defined heuristic